File: datasets/python_source-code/Multiple-A3/organic_3.py
# Linked List Reversal - organic submission 3
# Class-based, but different naming/API: insert_end, reverse_list, print_list.

class Node:
    def __init__(self, val):
        self.val = val
        self.next_node = None


class SinglyLinkedList:
    def __init__(self):
        self.first = None

    def insert_end(self, val):
        n = Node(val)
        if self.first is None:
            self.first = n
            return
        walker = self.first
        count = 0
        while walker.next_node is not None:
            walker = walker.next_node
            count += 1
        walker.next_node = n

    def reverse_list(self):
        back = None
        here = self.first
        for _ in range(self._length()):
            forward = here.next_node
            here.next_node = back
            back = here
            here = forward
        self.first = back

    def _length(self):
        c = 0
        walker = self.first
        while walker:
            c += 1
            walker = walker.next_node
        return c

    def print_list(self):
        vals = []
        walker = self.first
        while walker:
            vals.append(walker.val)
            walker = walker.next_node
        print(vals)


if __name__ == "__main__":
    sl = SinglyLinkedList()
    for x in (7, 14, 21, 28):
        sl.insert_end(x)
    sl.print_list()
    sl.reverse_list()
    sl.print_list()

File: server/datasets/java_source-code/Multiple-A3/organic_3.java
// Linked List Reversal - organic submission 3
// Class-based, naming/API: insertEnd, reverseList, printList.

class Node {
    int val;
    Node nextNode;

    public Node(int val) {
        this.val = val;
        this.nextNode = null;
    }
}

class SinglyLinkedList {
    Node first;

    public SinglyLinkedList() {
        this.first = null;
    }

    public void insertEnd(int val) {
        Node node = new Node(val);
        if (this.first == null) {
            this.first = node;
            return;
        }
        Node temp = this.first;
        while (temp.nextNode != null) {
            temp = temp.nextNode;
        }
        temp.nextNode = node;
    }

    public void reverseList() {
        Node left = null;
        Node mid = this.first;
        while (mid != null) {
            Node right = mid.nextNode;
            mid.nextNode = left;
            left = mid;
            mid = right;
        }
        this.first = left;
    }

    public void printList() {
        Node p = this.first;
        StringBuilder sb = new StringBuilder();
        while (p != null) {
            sb.append(p.val);
            if (p.nextNode != null) sb.append(" -> ");
            p = p.nextNode;
        }
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        SinglyLinkedList sll = new SinglyLinkedList();
        for (int v : new int[]{7, 14, 21, 28}) {
            sll.insertEnd(v);
        }
        sll.printList();
        sll.reverseList();
        sll.printList();
    }
}
